### PR TITLE
New Child fields for active status

### PR DESCRIPTION
**Pull request by wittejm**
# 💅🏼 What issue does this fix?

#1189
## 🍂 Before You Submit

* [x] Did you write tests? 
Just one: for new scope `not_deleted` on child. All other changes handled by updates to other tests.
* [x] Did you run Google Lighthouse and/or WebAIM (Wave) on UI components in your PR? **N/A**
* [x] Does your PR contain any required translations?  **no**
* [x] Did you run rake prep?
* [x] Are your primary keys UUIDs on any new tables?  **N/A**

## 🛷 Deployment

* [ ] Data Migrations
* [x] Schema Migrations
* [ ] Dependencies

## 🧳 Steps to test

Test child PUT endpoint in postman
## 🕯 Caveats, concerns

new date field is `last_active_date` instead of the `inactive_date` listed in the issue #1189

### CODE REVIEW

File: app/models/child.rb [hunk 1]
```diff
@@ -20,10 +20,19 @@ class Child < UuidApplicationRecord
   # this will try to validate with the incorrect dob even though the record has already failed
   validates :full_name, uniqueness: { scope: %i[date_of_birth business_id] }, unless: -> { errors }
 
+  REASONS = %w[
+    no_longer_in_my_care
+    no_longer_recieving_subsidies
+    other
+  ].freeze
+  validates :inactive_reason, inclusion: { in: REASONS }, allow_nil: true
```
Comment (katelovescode) [inline app/models/child.rb:28]:
```suggestion

  validates :inactive_reason, inclusion: { in: REASONS }, allow_nil: true
```

Comment (katelovescode) [reply app/models/child.rb:28]: Just to add a space for readability

File: spec/models/child_spec.rb [hunk 2]
```diff
@@ -56,6 +56,7 @@
 
   context 'scopes' do
     let(:inactive_child) { create(:child, active: false) }
+    let(:deleted_child) { create(:child, deleted: true) }
 
     it 'only displays active children in the active scope' do
       expect(Child.active).to include(child)
```
Comment (katelovescode) [inline spec/models/child_spec.rb:62]:
```suggestion
      expect(Child.active).to include(child)
      expect(Child.active).to not_include(deleted_child)
      expect(Child.active).to not_include(inactive_child)
```

File: spec/models/child_spec.rb [hunk 3]
```diff
@@ -66,6 +67,11 @@
       expect(Child.approved_for_date(child.approvals.first.effective_on)).to include(child)
       expect(Child.approved_for_date(child.approvals.first.effective_on - 1.day)).to eq([])
```
Comment (katelovescode) [inline spec/models/child_spec.rb:68]:
```suggestion
      expect(Child.approved_for_date(child.approvals.first.effective_on)).to not_include(inactive_child)
      expect(Child.approved_for_date(child.approvals.first.effective_on)).to not_include(deleted_child)
```

File: spec/models/child_spec.rb [hunk 4]
```diff
@@ -66,6 +67,11 @@
...
     end
+
+    it 'displays inactive children but not deleted children in the not_deleted scope' do
+      expect(Child.not_deleted).to include(inactive_child)
```
Comment (katelovescode) [inline spec/models/child_spec.rb:72]:
```suggestion
      expect(Child.not_deleted).to include(child)
      expect(Child.not_deleted).to include(inactive_child)
```

File: spec/models/child_spec.rb [hunk 5]
```diff
@@ -56,6 +56,7 @@
```
Comment (katelovescode) [inline spec/models/child_spec.rb:56]: Along with the validation specs above (line 13 - 48) can we add a validation spec for the inclusion validator, for example:
```
expect(subject).to validate_inclusion_of(:active).in_array(%w[Y N])
```

Also could you duplicate the date_param `it` block for `last_active_date`

**Review by katelovescode — CHANGES_REQUESTED** [review]
Looks really great, just a couple notes!

File: spec/models/child_spec.rb (hunk 2 above)
Comment (wittejm) [reply spec/models/child_spec.rb:109]: I think deleted entries should be included in any scope including `active`, unless the `not_deleted` scope is explicit. One alternative would be to adjust `active` scope to be `where(active: true, deleted: false)` , which seems not great, or to more generally apply `not_deleted` as a default scope everywhere and provide an inversion `include_deleted`. I think the hassle of adding `not_deleted` everywhere is worth it for avoiding the possible confusion. Also it looks like default scopes are frowned upon: https://piechowski.io/post/why-is-default-scope-bad-rails/

File: spec/models/child_spec.rb (hunk 3 above)
Comment (wittejm) [reply spec/models/child_spec.rb:116]: Some thing here: inactive and deleted entries should still be present IMO.

**Comment by wittejm** [comment]
Good idea re new validation tests. I discovered that 

`validates :last_active_date, date_param: true, allow_nil: true`

actually passes on an invalid string, because the string is quietly cast to a `nil`. https://stackoverflow.com/a/35553281

So I tweaked the validation rule to check for nil before the type cast. 

My comments above address your other suggested changed.

File: app/models/child.rb [hunk 6]
```diff
@@ -20,10 +20,20 @@ class Child < UuidApplicationRecord
...
   validates :full_name, uniqueness: { scope: %i[date_of_birth business_id] }, unless: -> { errors }
 
+  REASONS = %w[
+    no_longer_in_my_care
+    no_longer_recieving_subsidies
+    other
+  ].freeze
+
+  validates :inactive_reason, inclusion: { in: REASONS }, allow_nil: true
+  validates :last_active_date, date_param: true, unless: proc { |c| c.last_active_date_before_type_cast.nil? }
```
Comment (katelovescode) [inline app/models/child.rb:30]: TIL. 😳😳😳 This is awesome.

**Review by katelovescode — APPROVED** [review]

**Pull request merged**